File: Sort, Search, Tree1(BFT, DFT, BFS, DFS)/leetcode/404_sum_of_left_leaves/404_sum_of_left_leaves.c
/*
C에서는 Queue 구현이 없기 때문에, 
Queue Struct을 별도로 구현해야 한다.
*/

#include <stdio.h>
#include <stdlib.h>

typedef struct TreeNode{
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
} TreeNode;


TreeNode* createNode(int val){
    TreeNode* newNode = (TreeNode*)malloc(sizeof(TreeNode));
    newNode ->val = val;
    newNode ->left = NULL;
    newNode ->right = NULL;
    return newNode;
}

typedef struct QueueNode{
    TreeNode* treeNode;
    struct QueueNode* next;
} QueueNode;

typedef struct {
    QueueNode *front, *rear;
} Queue;

Queue* createQueue() {
    Queue* q = (Queue*)malloc(sizeof(Queue));
    q->front = q->rear = NULL;
    return q;
}


// Queue에 원소 추가
void enqueue(Queue* q, TreeNode* node) {
    QueueNode* temp = (QueueNode*)malloc(sizeof(QueueNode));
    temp->treeNode = node;
    temp->next = NULL;
    if (q->rear == NULL) {
        q->front = q->rear = temp;
        return;
    }
    q->rear->next = temp;
    q->rear = temp;
}

// Queue에서 원소 제거 및 반환
TreeNode* dequeue(Queue* q) {
    if (q->front == NULL) return NULL;
    QueueNode* temp = q->front;
    TreeNode* node = temp->treeNode;
    q->front = q->front->next;
    if (q->front == NULL) q->rear = NULL;
    free(temp);
    return node;
}

// Queue가 비었는지 확인
int isEmpty(Queue* q) {
    return q->front == NULL;
}



int sumOfLeftLeaves(TreeNode* root) {
    if (root==NULL) return 0;

    int ans = 0;
    Queue* q  = createQueue();
    enqueue(q, root);


    while(!isEmpty(q)){
        TreeNode* curnode = dequeue(q);;

        if (curnode->left !=NULL){
            TreeNode* temp = curnode -> left;
            if (temp->left == NULL && temp->right == NULL){
                ans += temp->val;

            }
            else{
                enqueue(q, temp);
            }
        }

        if (curnode-> right !=NULL){
            enqueue(q, curnode->right);
        }
    }

    return ans;
}

File: Sort, Search, Tree1(BFT, DFT, BFS, DFS)/leetcode/404_sum_of_left_leaves/404_sum_of_left_leaves.py
# Definition for a binary tree node.


'''
return sum of all left leaves
left leaf : leaf that is the left child of another node (with no child)
- no children tree node
- 어떤 노드의 left child 이면서, children이 아무것도 없을 때.



Input: root = [3,9,20,null,null,15,7] : BFT ordering
Output: 24
Explanation: There are two left leaves in the binary tree, with values 9 and 15 respectively.


Example 2:

Input: root = [1] : root만 있을 경우는 leaf가 아님.
Output: 0

접근 방식
- 
'''



class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


from typing import Optional

class Solution:
    def sumOfLeftLeaves(self, root: Optional[TreeNode]) -> int:
        

        ans = 0
        q = [root]
        
        while q:
            cur = q.pop(0)

            if cur.left is not None:
                

                temp = cur.left
                if (temp.left is None) and (temp.right is None): #left leaf
                    ans += temp.val
                
                
                else: # child가 있을 때 traversal할 이유가 있기에. leaf node는 필요없다.
                    q.append(cur.left)


            if cur.right is not None:
                q.append(cur.right)


        return ans



### 번외
def buils_tree(node_vals):
    if not node_vals:
        return None
    
    root =TreeNode(node_vals[0])
    q = [root]
    i  = 1 # indexing and counting

    while q and i<len(node_vals):
        node = q.pop(0)

        if node_vals[i] is not None:
            node.left = TreeNode(node_vals[i])
            q.append(node.left)
        i+=1

        if i<len(node_vals) and node_vals[i] is not None:
            node.right = TreeNode(node_vals[i])
            q.append(node.right)
        i+=1

    return root




def main():

    node_vals = [3,9, 20, None, None, 15, 7]

    root =buils_tree(node_vals)

    Sol = Solution()
    result = Sol.sumOfLeftLeaves(root)
    print("Ans : ", result)


if __name__ =="__main__":
    main()
